torch/lib/c10d/ProcessGroupMPI.cpp

```diff
@@ -588,6 +588,167 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::reduce_scatter(
+      [opts, this, inputSplitSizes, outputSplitSizes](std::unique_ptr<WorkEntry>& entry) {
+        auto srcdata = (entry->src)[0];
+        auto dstdata = (entry->dst)[0];
+        std::vector<int> send_lengths(size_);
+        std::vector<int> recv_lengths(size_);
+        std::vector<int> send_offsets(size_);
+        std::vector<int> recv_offsets(size_);
+        auto computeLengthsAndOffsets = [this](std::vector<int> split_sizes, at::Tensor &tensor, std::vector<int> &lengths, std::vector<int> &offsets) {
+          if(split_sizes.size() == 0) {
+            int split_size = tensor.numel() / size_;
+            for(int i = 0; i < size_; i++) {
+              lengths[i] = split_size;
+              offsets[i] = i * split_size;
+            }
+          } else {
+            int split_size = tensor.numel() / tensor.size(0);
+            int offset = 0;
+            for(int i = 0; i < size_; i++) {
+              int length = split_size * split_sizes[i];
+              lengths[i] = length;
+              offsets[i] = offset;
+              offset += length;
+            }
+          }
+        };
+        computeLengthsAndOffsets(inputSplitSizes, srcdata, send_lengths, send_offsets);
+        computeLengthsAndOffsets(outputSplitSizes, dstdata, recv_lengths, recv_offsets);
+        c10::DeviceGuard guard(srcdata.device());
+        std::unique_lock<std::mutex> globalLock(pgGlobalMutex_);
+        MPI_CHECK(MPI_Alltoallv(
+              srcdata.data_ptr(),
+              send_lengths.data(),
+              send_offsets.data(),
+              mpiDatatype.at(srcdata.scalar_type()),
+              dstdata.data_ptr(),
+              recv_lengths.data(),
+              recv_offsets.data(),
+              mpiDatatype.at(dstdata.scalar_type()),
+              pgComm_));
+      };
+    auto entry = std::unique_ptr<WorkEntry>(
+        new WorkEntry(&inputTensors, &outputTensors, std::move(runFunc)));
+    return enqueue(std::move(entry));
+  }
+}
+std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::alltoall(
+    std::vector<std::vector<at::Tensor>>& outputTensors,
+    std::vector<std::vector<at::Tensor>>& inputTensors,
+    const AllToAllOptions& opts) {
+  if(inputTensors[0].size() != size_ || outputTensors[0].size() != size_) {
+    throw std::runtime_error("Number of input or output tensors are not equal to group size");
+  }
+  std::function<void(std::unique_ptr<WorkEntry>&)> runFunc =
+    [opts, this](std::unique_ptr<WorkEntry>& entry) {
+      std::vector<int> send_lengths(size_);
+      std::vector<int> recv_lengths(size_);
+      std::vector<int> send_offsets(size_);
+      std::vector<int> recv_offsets(size_);
+      auto srcdata = entry->src;
+      auto dstdata = entry->dst;
+      auto computeLengthsAndOffsets = [this](std::vector<at::Tensor> &tensors, std::vector<int> &lengths, std::vector<int> &offsets) {
+        int offset = 0;
+        for(int i = 0; i < size_; i++) {
+          lengths[i] = tensors[i].numel();
```

Comment:
Similar to the comment above, this is doing an unchecked cast from `int64_t` to `int` which can overflow for tensors > 2GB.

Reply:
As mentioned above, will fix it